Add in persistentshare model and views and hook it into the target sh…

jnation3406 opened the pull request:
…are page.

There are so many urls added since I added both API views with django rest framework, and template / management views for normal tom users. The template views use the DRF views to actually create/delete persistentshare instances. There is also a version of URLS that applies to all targets, vs one filtered on a specific target for both creation and management. I tried to get permissions such that only a user with target_share permissions and permissions on that specific target should be able to create a persistent share on that target.

[manage persistent shares](https://github.com/user-attachments/assets/dceb0428-a954-4e24-b2bc-ad2172b3f5db)

jchate6 commented on tom_targets/templates/tom_targets/target_manage_persistent_shares.html:
```diff
@@ -0,0 +1,22 @@
+<script src="https://cdn.jsdelivr.net/npm/js-base64@3.7.5/base64.min.js"></script>
+{% endblock %}
+{% block content %}
+<h3>Create Persistent Share {% if target %} for Target <a href="{% url 'targets:detail' pk=target.id %}" title="Back">{{ target.name }}</a>{% endif %}</h3>
```
This needs more explanation and possibly a link to documentation. What will this actually do?

"Persistent share" is not a term that would be obvious or familiar to an Astronomer or TOM admin. 
I think generally I would stay away from referring to a "Persistent Share". 
First, I think I'd rename the concept to "Continuous Sharing", then change this line to `<h3>Continuously Share {% if target %} data for <a href="{% url 'targets:detail' pk=target.id %}" title="Back">{{ target.name }}</a>{% endif %}</h3>`

jnation3406 replied:
Happy to change the verbiage to Continuously Share, and I'll add a section to the sharing help pages describing this.

jchate6 commented on tom_targets/templates/tom_targets/target_manage_persistent_shares.html:
```diff
@@ -0,0 +1,22 @@
+<div id='target-persistent-share-create'>
+  {% create_persistent_share target %}
+</div>
+<h3>Manage Persistent Shares {% if target %} for Target <a href="{% url 'targets:detail' pk=target.id %}" title="Back">{{ target.name }}</a>{% endif %}</h3>
```
"Manage Continuous Sharing for"

jchate6 commented on tom_targets/templates/tom_targets/partials/persistent_share_table.html:
```diff
@@ -0,0 +1,52 @@
+      table.innerHTML = html;
+    })
+  }
+</script>
```
No new line

jchate6 reviewed (changes requested):
Only part of the way through the review. Will try to finish tomorrow.

jnation3406 commented on tom_targets/views.py:
```diff
@@ -885,3 +889,33 @@ def post(self, request, *args, **kwargs):
+
+
+class TargetPersistentShareManageFormView(PersistentShareManageFormView):
+    def get(self, request, target_pk):
```
I'm not sure what you mean here... This subclasses the PersistentShareManageFormView which doesn't have a target_pk in the get because its mapped to a url with no target params. It is just using the class stuff from that base class but reimplementing the get method to accept the target_pk since this class is mapped to a url with the target_pk. So this one always expects a target_pk, and the base class never expects a target_pk.

jchate6 commented on tom_targets/forms.py:
```diff
@@ -241,3 +241,32 @@ class TargetMergeForm(forms.Form):
+        if self.target_id:
+            self.fields['target'].queryset = Target.objects.filter(pk=self.target_id)
+        else:
+            if self.user:
```
I feel like there is a permissions issue here.
Without any share permissions I can still add a persistent share from the target detail sharing page.
I can also still access the `persistentshare/manage/` page and delete any of the persistent shares.
I cannot however add new persistent shares from the dropdown.

Also, a dropdown of all targets in a TOM (which is what any superuser would see from here) will be effectively useless. (Remember Seguaro TOM as 7+ million targets.)

jnation3406 replied:
Are you saying that your user has no `target_share` permissions but is still able to create/destroy persistent shares? Also, what do you suggest we provide instead of a dropdown?

jnation3406 replied:
hmm I'll try to look into that. That is certainly strange since I can always delete persistent shares. My intent was that only permissioned users can create/delete them, but the permissions can be both on individual targets and on the `target_share` permission.

jchate6 reviewed (changes requested):
Some more things...

jnation3406 commented on tom_targets/base_models.py:
```diff
@@ -403,6 +403,7 @@ class Meta:
         permissions = (
             ('view_target', 'View Target'),
             ('add_target', 'Add Target'),
+            ('share_target', 'Share Target'),
```
You actually added these permissions to views: https://github.com/TOMToolkit/tom_base/commit/a11fee132b681a1d4502768200c16665791f152e. You just never actually "created" the permissions, so checking the permissions was breaking things. I am just attempting to follow what you were doing.

jnation3406 commented on tom_targets/signals/handlers.py:
```diff
@@ -0,0 +1,13 @@
+
+
+@receiver(post_save, sender=ReducedDatum)
+def cb_dataproduct_post_save(sender, instance, *args, **kwargs):
```
cb just refers to callback, and I think is just the notation used in django examples for these. We could call it whatever I think.

jnation3406 commented on tom_targets/templates/tom_targets/target_manage_persistent_shares.html:
This template is a dedicated manage form and is linked at `/targets/persistentshare/manage/` and `/targets/{target_pk}/persistentshare/manage/`

jchate6 replied:
I mean, how would a user of a standard TOM get here without typing the path into their browser?

jchate6 replied:
Is that not an expected use case?

jnation3406 replied:
It's there if someone wants to navigate to it, but its linked to any page by default. In the snex2 TOM, I include the partials into the custom target_details page. But I still navigate to that page for testing, and I could imagine a custom TOM linking to it somewhere.

jnation3406 commented on tom_dataproducts/data_processor.py:
```diff
@@ -70,7 +71,14 @@ def run_data_processor(dp):
-    ReducedDatum.objects.bulk_create(new_reduced_datums)
+    reduced_datums = ReducedDatum.objects.bulk_create(new_reduced_datums)
+
+    # 4. Trigger any sharing you may have set to occur when new data comes in
```
There is also a signal handler on ReducedDatum post_save which should work if Reduced Datums are individually created. If there is some other avenues for bulk_create of ReducedDatums then that would not work right now.

jnation3406 commented:
I re-did the permissions and removed the `share_target` permission everywhere. Now its all using the `change_target` permissions + the persistentshare permissions. I also updated the pages to show/hide stuff based on your users permissions. 

This is when you only have view_persistentshare permissions + target permissions
![image](https://github.com/user-attachments/assets/d2553b56-5edf-46db-b601-ca9c1a3f5a28)

This is view + delete:
![image](https://github.com/user-attachments/assets/5bbe1a85-5fee-4924-8fdc-67da11eddce3)

And this is view + delete + add (basically all permissions, change isn't used since they can't be modified):
![image](https://github.com/user-attachments/assets/a3625fde-acf3-4d0b-a992-7e3657119c89)

jchate6 commented:
The permissions changes seem to work really well.
Will deleting the user who made a persistent share have any unforeseen repercussions? It doesn't look like there are any immediate problems, but I haven't tested actually triggering a share.

jnation3406 commented:
> The permissions changes seem to work really well. Will deleting the user who made a persistent share have any unforeseen repercussions? It doesn't look like there are any immediate problems, but I haven't tested actually triggering a share.

It shouldn't have any effect. Anyone with permissions can see/delete any persistent shares, the user that created it is just for knowledge purposes to know who you can ask for why they did that in your collaboration.

jchate6 commented:
I get an error when I try to add a persistent share via the admin interface:
```TypeError: BaseModelForm.__init__() got an unexpected keyword argument 'user'```

jnation3406 commented:
> I get an error when I try to add a persistent share via the admin interface: `TypeError: BaseModelForm.__init__() got an unexpected keyword argument 'user'`

Oops I had broken that when changing the forms before. It's fixed now.

jchate6 reviewed (approved):
Looks good!
